### PR TITLE
[New Resource] `azurerm_machine_learning_workspace_network_outbound_rule_private_endpoint`

### PR DESCRIPTION
**Pull request by xuzhang3**
## Community Note

* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review

## Description

Add support for Machine Learning Workspace Private Endpoint Network Outbound Rule :
- [ ] FQDN (`azurerm_machine_learning_workspace_network_outbound_rule_fqdn`)
- [x] Private Endpoint
- [ ] Service Tag 

```log
=== RUN   TestAccMachineLearningWorkspaceNetworkOutboundRulePrivateEndpoint_onlyApprovedOutbound
=== PAUSE TestAccMachineLearningWorkspaceNetworkOutboundRulePrivateEndpoint_onlyApprovedOutbound
=== RUN   TestAccMachineLearningWorkspaceNetworkOutboundRulePrivateEndpoint_internetOutbound
=== PAUSE TestAccMachineLearningWorkspaceNetworkOutboundRulePrivateEndpoint_internetOutbound
=== RUN   TestAccMachineLearningWorkspaceNetworkOutboundRulePrivateEndpoint_keyVault
=== PAUSE TestAccMachineLearningWorkspaceNetworkOutboundRulePrivateEndpoint_keyVault
=== RUN   TestAccMachineLearningWorkspaceNetworkOutboundRulePrivateEndpoint_workspace
=== PAUSE TestAccMachineLearningWorkspaceNetworkOutboundRulePrivateEndpoint_workspace
=== RUN   TestAccMachineLearningWorkspaceNetworkOutboundRulePrivateEndpoint_redis
=== PAUSE TestAccMachineLearningWorkspaceNetworkOutboundRulePrivateEndpoint_redis
=== RUN   TestAccMachineLearningWorkspaceNetworkOutboundRulePrivateEndpoint_requiresImport
=== PAUSE TestAccMachineLearningWorkspaceNetworkOutboundRulePrivateEndpoint_requiresImport
=== CONT  TestAccMachineLearningWorkspaceNetworkOutboundRulePrivateEndpoint_onlyApprovedOutbound
=== CONT  TestAccMachineLearningWorkspaceNetworkOutboundRulePrivateEndpoint_requiresImport
=== CONT  TestAccMachineLearningWorkspaceNetworkOutboundRulePrivateEndpoint_redis
=== CONT  TestAccMachineLearningWorkspaceNetworkOutboundRulePrivateEndpoint_internetOutbound
=== CONT  TestAccMachineLearningWorkspaceNetworkOutboundRulePrivateEndpoint_workspace
=== CONT  TestAccMachineLearningWorkspaceNetworkOutboundRulePrivateEndpoint_keyVault
--- PASS: TestAccMachineLearningWorkspaceNetworkOutboundRulePrivateEndpoint_requiresImport (480.79s)
--- PASS: TestAccMachineLearningWorkspaceNetworkOutboundRulePrivateEndpoint_workspace (552.37s)
--- PASS: TestAccMachineLearningWorkspaceNetworkOutboundRulePrivateEndpoint_onlyApprovedOutbound (558.66s)
--- PASS: TestAccMachineLearningWorkspaceNetworkOutboundRulePrivateEndpoint_keyVault (561.82s)
--- PASS: TestAccMachineLearningWorkspaceNetworkOutboundRulePrivateEndpoint_internetOutbound (594.98s)
--- PASS: TestAccMachineLearningWorkspaceNetworkOutboundRulePrivateEndpoint_redis (1446.73s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/machinelearning       1464.087s
```

## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”

## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevent documentation.
- [x] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [ ] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.

## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

* `azurerm_resource` - support for the `thing1` property [GH-00000]

This is a (please select all that apply):

- [ ] Bug Fix
- [x] New Feature (ie adding a service, resource, or data source)
- [ ] Enhancement
- [ ] Breaking Change

## Related Issue(s)
Fixes #0000

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

### CODE REVIEW

**Comment by xuzhang3** [comment]
CI error is not relate with this PR. All tests passed

File: website/docs/r/machine_learning_workspace_network_outbound_rule_private_endpoint.html.markdown [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+* `id` - The ID of the Machine Learning Workspace Network Outbound Rule Private Endpoint.
+
+### Timeouts
+
+The `timeouts` block allows you to
+specify [timeouts](https://www.terraform.io/language/resources/syntax#operation-timeouts) for certain actions:
+
+* `create` - (Defaults to 30 minutes) Used when creating the Machine Learning Workspace Network Outbound Rule Private Endpoint.
+* `update` - (Defaults to 30 minutes) Used when updating the Machine Learning Workspace Network Outbound Rule Private Endpoint.
```
Comment (catriona-m) [inline website/docs/r/machine_learning_workspace_network_outbound_rule_private_endpoint.html.markdown:118]: this resource doesn't appear to have an update func?

**Review by catriona-m — CHANGES_REQUESTED** [review]
Hi @xuzhang3, I had a look through this and left some comments inline, once those are addressed I can take another look. Thanks!

File: internal/services/machinelearning/machine_learning_workspace_network_outbound_rule_private_endpoint_resource.go [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+			Required:     true,
+			ForceNew:     true,
+			ValidateFunc: managednetwork.ValidateWorkspaceID,
+		},
+
+		"service_resource_id": {
+			Type:         pluginsdk.TypeString,
+			Required:     true,
+			ForceNew:     true,
+			ValidateFunc: validation.StringIsNotEmpty,
```
[the remark this answers is not in the document]
Comment (xuzhang3) [reply internal/services/machinelearning/machine_learning_workspace_network_outbound_rule_private_endpoint_resource.go:73]: The service ID varies based on the service resource type. I intentionally only check the empty string to ensure that the property may introduce resources that may not follow the pattern of `/subscriptions/<sub ID>/resourceGroups/<resource group name>/...`

File: internal/services/machinelearning/machine_learning_workspace_network_outbound_rule_private_endpoint_resource.go [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+		Timeout: 30 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			client := metadata.Client.MachineLearning.ManagedNetwork
+
+			id, err := managednetwork.ParseOutboundRuleID(metadata.ResourceData.Id())
+			if err != nil {
+				return err
+			}
+
+			future, err := client.SettingsRuleDelete(ctx, *id)
```
[the remark this answers is not in the document]
Comment (xuzhang3) [reply internal/services/machinelearning/machine_learning_workspace_network_outbound_rule_private_endpoint_resource.go]: Has been updated to `SettingsRuleDeleteThenPoll`

File: website/docs/r/machine_learning_workspace_network_outbound_rule_private_endpoint.html.markdown (hunk 1 above)
Comment (xuzhang3) [reply website/docs/r/machine_learning_workspace_network_outbound_rule_private_endpoint.html.markdown:118]: Yes, this resource cannot be updated. I follow the behavior in the portal.

**Comment by xuzhang3** [comment]
> this and left some comments inline, once those are addressed I can take another look. Thanks!

Thank for reviewing. updated as suggested.

File: internal/services/machinelearning/machine_learning_workspace_network_outbound_rule_private_endpoint_resource.go [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+				Name:        id.OutboundRuleName,
+				WorkspaceId: managednetwork.NewWorkspaceID(id.SubscriptionId, id.ResourceGroupName, id.WorkspaceName).ID(),
+			}
+
+			if model := resp.Model; model != nil {
+				if props := model.Properties; props != nil {
+					if prop, ok := props.(managednetwork.PrivateEndpointOutboundRule); ok && prop.Destination != nil {
+						if prop.Destination.SparkEnabled != nil {
+							state.SparkEnabled = *prop.Destination.SparkEnabled
+						}
```
Comment (catriona-m) [inline internal/services/machinelearning/machine_learning_workspace_network_outbound_rule_private_endpoint_resource.go:208]:
```suggestion
						state.SparkEnabled = pointer.From(prop.Destination.SparkEnabled)
```

File: internal/services/machinelearning/machine_learning_workspace_network_outbound_rule_private_endpoint_resource.go [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+
+						if prop.Destination.SubresourceTarget != nil {
+							state.SubresourceTarget = *prop.Destination.SubresourceTarget
+						}
```
Comment (catriona-m) [inline internal/services/machinelearning/machine_learning_workspace_network_outbound_rule_private_endpoint_resource.go:212]:
```suggestion
						state.SubresourceTarget = pointer.From(prop.Destination.SubresourceTarget)
```

File: internal/services/machinelearning/machine_learning_workspace_network_outbound_rule_private_endpoint_resource.go [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+
+						if prop.Destination.ServiceResourceId != nil {
+							state.ServiceResourceId = *prop.Destination.ServiceResourceId
+						}
```
Comment (catriona-m) [inline internal/services/machinelearning/machine_learning_workspace_network_outbound_rule_private_endpoint_resource.go:216]:
```suggestion
						state.ServiceResourceId = pointer.From(prop.Destination.ServiceResourceId)
```

File: internal/services/machinelearning/machine_learning_workspace_network_outbound_rule_private_endpoint_resource.go [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+			}
+
+			resId, err := resourceids.ParseAzureResourceID(model.ServiceResourceId)
+			if err != nil {
+				return err
+			}
+
+			supportType := false
+			if subTypes, ok := resourceTypeSupportSubResType[resId.Provider]; ok {
+				for _, typ := range subTypes {
```
Comment (catriona-m) [inline internal/services/machinelearning/machine_learning_workspace_network_outbound_rule_private_endpoint_resource.go:141]:
```suggestion
				for _, type := range subTypes {
```

File: internal/services/machinelearning/machine_learning_workspace_network_outbound_rule_private_endpoint_resource.go [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+		},
+
+		"service_resource_id": {
+			Type:         pluginsdk.TypeString,
+			Required:     true,
+			ForceNew:     true,
+			ValidateFunc: validation.StringIsNotEmpty,
+		},
+
+		"sub_resource_target": {
```
Comment (catriona-m) [inline internal/services/machinelearning/machine_learning_workspace_network_outbound_rule_private_endpoint_resource.go:76]: is this something we would be able to infer based on the `service_resource_id`

File: website/docs/r/machine_learning_workspace_network_outbound_rule_private_endpoint.html.markdown (hunk 1 above)
Comment (catriona-m) [reply website/docs/r/machine_learning_workspace_network_outbound_rule_private_endpoint.html.markdown:118]: could we remove the `update` timeout here?

**Review by catriona-m — CHANGES_REQUESTED** [review]
Thanks for updating this @xuzhang3 - I left a couple more comments inline but once those are addressed I can take another look. Thanks!

File: internal/services/machinelearning/machine_learning_workspace_network_outbound_rule_private_endpoint_resource.go (hunk 7 above)
Comment (xuzhang3) [reply internal/services/machinelearning/machine_learning_workspace_network_outbound_rule_private_endpoint_resource.go:141]: `type` is the in-built type of go, cannot name to type

Comment (catriona-m) [reply internal/services/machinelearning/machine_learning_workspace_network_outbound_rule_private_endpoint_resource.go:141]: oops! right you are.

**Review by catriona-m — APPROVED** [review]
Thanks @xuzhang3 LGTM!

**Pull request merged**